Use Partial for the assign/mixin typings

The sources for assign/mixin functions should be typed so that subsequent sources can contain attributes that are not included in the first `source`. At the moment compile errors occurs if the second source expands the keys compared to the initial `source`.

Code:
```ts
const baseObject = {
	foo: 'bar'
};

lang.assign({}, baseObject, { baz: 'foo' });
```

Expected:

Compile successfully

Actual:

`error TS2453: The type argument for type parameter 'U' cannot be inferred from the usage. Consider specifying the type arguments explicitly.
  Type argument candidate '{ foo: string; }' is not a valid type argument because it is not a supertype of candidate '{ baz: string; }'.
    Object literal may only specify known properties, and 'baz' does not exist in type '{ foo: string; }'.`

--- a/src/lang.ts
+++ b/src/lang.ts
@@ -87,7 +87,10 @@ function _mixin<T extends {}, U extends {}>(kwArgs: MixinArgs<T, U>): T&U {
 }
 
 interface ObjectAssignConstructor extends ObjectConstructor {
-	assign<T extends {}, U extends {}>(target: T, ...sources: (U | null | undefined)[]): T&U;
+	assign<T, U>(target: T, source: U): T & U;
+	assign<T, U1, U2>(target: T, source1: U1, source2: U2): T & U1 & U2;
+	assign<T, U1, U2, U3>(target: T, source1: U1, source2: U2, source3: U3): T & U1 & U2 & U3;
+	assign(target: any, ...sources: any[]): any;
 }
 
 /**
@@ -99,7 +102,7 @@ interface ObjectAssignConstructor extends ObjectConstructor {
  */
 export const assign = has('object-assign') ?
 	(<ObjectAssignConstructor> Object).assign :
-	function<T extends {}, U extends {}> (target: T, ...sources: (U | null | undefined)[]): T&U {
+	function (target: any, ...sources: any[]): any {
 		return _mixin({
 			deep: false,
 			inherited: false,

--- a/tests/unit/lang.ts
+++ b/tests/unit/lang.ts
@@ -57,7 +57,7 @@ registerSuite({
 			property2: 'value2'
 		};
 
-		lang.assign<typeof object, typeof source1 | typeof source3>(object, source1, <any> null, source3);
+		lang.assign(object, source1, <any> null, source3);
 
 		assert.deepEqual(object, {
 			property1: 'value1',
@@ -97,6 +97,32 @@ registerSuite({
 		const alsoAssigned: {} & ({ a: number, b: number } | { c: number, d: number }) = lang.assign(object, source1, source2, source3);
 
 		assert(alsoAssigned);
+	},
+
+	'.assign() with different types of sources'() {
+		const baseObject = {
+			foo: 'foo'
+		};
+
+		const assignedObject = lang.assign({}, baseObject, { bar: 'bar' });
+		assert.strictEqual(assignedObject.bar, 'bar');
+		assert.strictEqual(assignedObject.foo, 'foo');
+	},
+	'.assign() with a source whose type is a subset of another'() {
+		type FooBar = {
+			foo: string;
+			bar: string;
+		};
+		const foobar = {
+			foo: 'foo',
+			bar: 'bar'
+		};
+		const bar = {
+			bar: 'baz'
+		};
+		const assignedObject = lang.assign({}, foobar, bar);
+		assert.strictEqual(assignedObject.foo, 'foo');
+		assert.strictEqual(assignedObject.bar, 'baz');
 	},
 
 	'.deepAssign()'() {

